cmake/DefaultTargetProperties.cmake

```diff
@@ -0,0 +1,129 @@
+
+
+    #-------------------------------------------------------------------
+    # Symbol visibility
+
+    # Hide symbols from this library by default.
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN YES
+    )
+
+    # Hide all symbols from external statically linked libraries.
+    if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_link_options(${target_name} PRIVATE -Wl,--exclude-libs,ALL)
+    endif ()
+
+    # Use C++11 ABI by default, but allow disabling.
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+        option(GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI" ON)
+        if (GLIBCXX_USE_CXX11_ABI)
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=1)
+        else ()
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=0)
+        endif ()
+    endif ()
+
+    # TODO(DF): POSITION_INDEPENDENT_CODE required for Python module? See OCIO.
+
+    #-------------------------------------------------------------------
+    # RPATH wrangling.
+
+    # With the 'usual' install path structure in mind, search from the
+    # bin directory (i.e. a binary loading a dynamic library) and then
+    # from the current directory (i.e. dynamic library loading another
+    # dynamic library).
+    if (APPLE)
+        set(rpath "@loader_path/../${CMAKE_INSTALL_LIBDIR};@loader_path")
+    else ()
+        set(rpath "$ORIGIN/../${CMAKE_INSTALL_LIBDIR};$ORIGIN")
+    endif ()
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Allow CMake to set a temporary runtime library search path in
+        # the build phase that potentially hardcodes absolute paths,
+        # useful for TDD.
+        SKIP_BUILD_RPATH FALSE
+        BUILD_WITH_INSTALL_RPATH FALSE
+        # Don't add hardcoded directories to final runtime search path.
+        INSTALL_RPATH_USE_LINK_PATH FALSE
+        # Runtime search path value
+        INSTALL_RPATH "${rpath}"
+        # Enable RPATH on OSX
+        MACOSX_RPATH ON
+    )
+
+    #-------------------------------------------------------------------
+    # C++ #define to toggle static vs dynamic lib
+
+    # Used by header generated by CMake's `generate_export_header`
+    if (NOT BUILD_SHARED_LIBS)
+        target_compile_definitions(${target_name} PRIVATE ${PROJECT_NAME}_STATIC_DEFINE=1)
+    endif ()
+
+
+    #-------------------------------------------------------------------
+    # Build tree directory structure.
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Make the build area layout look a bit more like the final dist
+        # layout
+        LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR}
```

Comment:
Are you sure you want it in CMAKE_BINARY_DIR and not CMAKE_CURRENT_BINARY_DIR?

Reply:
Hmm, the downside of `CMAKE_CURRENT_BINARY_DIR` is, AFAIK, that the outputs will be nested in a build directory structure that mirrors the source tree structure, rather than at the top-level of the build directory mirroring the install structure.

But I guess insisting on a top-level output would cause conflicts (or at least messiness) if this project was included as child of another project?

Perhaps this project's top-level CMakeLists could set e.g. a `OPENASSETIO_BIINARY_DIR=CMAKE_CURRENT_BINARY_DIR` variable that can be used by CMakeLists in subdirectories if they want a pseudo-top-level output?

Or perhaps we should just give up on a neat build directory and  use `CMAKE_CURRENT_BINARY_DIR` throughout.

Reply:
Aha, `PROJECT_BINARY_DIR` to the rescue?

Reply:
Switched to `PROJECT_BINARY_DIR` in 42817cf